Bookkeeping export fails for exports that take longer than 30 seconds

Error displayed in the frontend:
![Image](https://github.com/user-attachments/assets/4c56f864-2565-44c5-a807-c9f7edf3f2c5)

Traceback in the backend:

```
[2025-02-12 17:14:35 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:12)
[2025-02-12 17:14:35 +0000] [12] [ERROR] Error handling request /webshop/download-accounting-file/2024/10
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/gunicorn/workers/sync.py", line 134, in handle
    self.handle_request(listener, req, client, addr)
  File "/usr/local/lib/python3.11/site-packages/gunicorn/workers/sync.py", line 177, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 1536, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 1511, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 917, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/flask/app.py", line 902, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work/src/service/internal_service.py", line 90, in view_wrapper
    data = f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/work/src/shop/views.py", line 336, in download_accounting_file_route
    return b64encode(export_accounting(start_date, end_data, TimePeriod.Month, g.user_id).encode("cp437")).decode(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work/src/shop/accounting/export.py", line 60, in export_accounting
    completed_payments = get_completed_payments_from_stripe(start_date, end_date)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work/src/shop/completed_payment.py", line 98, in get_completed_payments_from_stripe
    return convert_completed_stripe_charges_to_payments(stripe_charges)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work/src/shop/completed_payment.py", line 54, in convert_completed_stripe_charges_to_payments
    intent = retry(lambda: PaymentIntent.retrieve(charge.payment_intent))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/work/src/shop/stripe_util.py", line 72, in retry
    return f()
           ^^^
  File "/work/src/shop/completed_payment.py", line 54, in <lambda>
    intent = retry(lambda: PaymentIntent.retrieve(charge.payment_intent))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stripe/_payment_intent.py", line 8343, in retrieve
    instance.refresh()
  File "/usr/local/lib/python3.11/site-packages/stripe/_api_resource.py", line 37, in refresh
    return self._request_and_refresh("get", self.instance_url())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stripe/_api_resource.py", line 110, in _request_and_refresh
    obj = StripeObject._request(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stripe/_stripe_object.py", line 366, in _request
    response, api_key = requestor.request(method_, url_, params, headers)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stripe/_api_requestor.py", line 96, in request
    rbody, rcode, rheaders, my_api_key = self.request_raw(
                                         ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stripe/_api_requestor.py", line 358, in request_raw
    rcontent, rcode, rheaders = self._client.request_with_retries(
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stripe/_http_client.py", line 141, in request_with_retries
    return self._request_with_retries_internal(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stripe/_http_client.py", line 168, in _request_with_retries_internal
    response = self.request(method, url, headers, post_data)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stripe/_http_client.py", line 326, in request
    return self._request_internal(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/stripe/_http_client.py", line 355, in _request_internal
    result = self._thread_local.session.request(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/requests/adapters.py", line 667, in send
    resp = conn.urlopen(
           ^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py", line 787, in urlopen
    response = self._make_request(
               ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py", line 534, in _make_request
    response = conn.getresponse()
               ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/urllib3/connection.py", line 516, in getresponse
    httplib_response = super().getresponse()
                       ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/http/client.py", line 1378, in getresponse
    response.begin()
  File "/usr/local/lib/python3.11/http/client.py", line 318, in begin
    version, status, reason = self._read_status()
                              ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/http/client.py", line 279, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/socket.py", line 706, in readinto
    return self._sock.recv_into(b)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/ssl.py", line 1311, in recv_into
    return self.read(nbytes, buffer)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/ssl.py", line 1167, in read
    return self._sslobj.read(len, buffer)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/gunicorn/workers/base.py", line 204, in handle_abort
    sys.exit(1)
SystemExit: 1
[2025-02-12 17:04:58 +0000] [15] [INFO] Worker exiting (pid: 15)
```

--- a/admin/src/Components/icons/uikit.tsx
+++ b/admin/src/Components/icons/uikit.tsx
@@ -8,6 +8,7 @@ export type UiKitIconName =
     | "chevron-up"
     | "cog"
     | "commenting"
+    | "cloud-download"
     | "eye-slash"
     | "home"
     | "lock"

--- a/admin/src/Sales/AccountingExport.jsx
+++ b/admin/src/Sales/AccountingExport.jsx
@@ -1,13 +1,18 @@
-import React, { useState } from "react";
+import React, { useEffect, useId, useRef, useState } from "react";
 import Select from "react-select";
-import { get } from "../gateway";
-import { showError, showSuccess } from "../message";
+import UIkit from "uikit";
+import Icon from "../Components/icons";
+import { get, post } from "../gateway";
+import Member from "../Models/Member";
 
 function create_option(label) {
     return { label: label, value: label };
 }
 
-function download_blob(blob, file_name) {
+function download_text(content, file_name) {
+    const blob = new Blob([content], {
+        type: "text/plain",
+    });
     const element = document.createElement("a");
     element.href = URL.createObjectURL(blob);
     element.download = file_name;
@@ -17,34 +22,31 @@ function download_blob(blob, file_name) {
     URL.revokeObjectURL(element.href);
 }
 
-function exportAccounting(year, month, file_name) {
-    if (file_name) {
-        file_name = file_name + ".si";
-    } else {
-        file_name = `Accounting_${year}_${month}.si`;
-    }
-
-    return get({
-        url: `/webshop/download-accounting-file/${year}/${month}`,
-    })
-        .then((response) => {
-            const blob = new Blob([atob(response.data)], {
-                type: "text/plain",
-            });
-            download_blob(blob, file_name);
-            showSuccess("Laddat ner SIE-fil för bokföring.");
-        })
-        .catch((error) => {
-            showError(
-                `<h2>Misslyckades ladda ner fil</h2>Kunde inte
-                        kommunicera med servern:
-                        <pre>${JSON.stringify(error.message)}</pre>
-                    </>`,
-            );
-        });
+function downloadAccounting(accounting_export) {
+    download_text(
+        accounting_export.content,
+        format_date(new Date(accounting_export.start_date)) +
+            "_" +
+            format_date(new Date(accounting_export.end_date)) +
+            "_" +
+            accounting_export.aggregation +
+            ".sie",
+    );
 }
 
-export default function AccountingExport() {
+function exportAccounting(year, month) {
+    return post({
+        url: `/webshop/accounting/export/${year}/${month}`,
+        expectedDataStatus: "ok",
+    }).then(({ data }) => {
+        UIkit.notification(`Export #${data} startad`, {
+            timeout: 1000,
+            status: "success",
+        });
+    });
+}
+
+function StartExportForm() {
     let year_options = [];
     let month_options = [];
     const current_year = new Date().getFullYear();
@@ -66,14 +68,287 @@ export default function AccountingExport() {
     const handleExportClick = (event) => {
         event.preventDefault();
         setLoading(true);
-        const promise = exportAccounting(
-            yearOption.label,
-            monthOption.label,
-            document.getElementById("file_name").value,
-        );
-        promise.then(() => setLoading(false));
+        const promise = exportAccounting(yearOption.label, monthOption.label);
+        promise.finally(() => setLoading(false));
     };
 
+    return (
+        <form className="uk-form-stacked">
+            <fieldset className="uk-margin-top">
+                <legend className="uk-legend">
+                    Välj vilken period du vill exportera
+                </legend>
+                <label className="uk-form-label" htmlFor="year">
+                    År:
+                </label>
+                <Select
+                    id="year"
+                    name="year"
+                    className="uk-width-1-1"
+                    tabIndex={1}
+                    options={year_options}
+                    value={yearOption}
+                    onChange={(from_year) => setYearOption(from_year)}
+                />
+                <label className="uk-form-label" htmlFor="month">
+                    Månad:
+                </label>
+                <Select
+                    id="month"
+                    name="month"
+                    className="uk-width-1-1"
+                    tabIndex={2}
+                    options={month_options}
+                    value={monthOption}
+                    onChange={(month) => setMonthOption(month)}
+                />
+                <button
+                    className="uk-button uk-button-primary uk-width-1-1"
+                    role="button"
+                    tabIndex={4}
+                    disabled={loading}
+                    onClick={handleExportClick}
+                >
+                    {loading ? "Laddar..." : "Starta export"}
+                </button>
+            </fieldset>
+        </form>
+    );
+}
+
+function status_to_emoji(status) {
+    if (status === "completed") {
+        return "✅";
+    }
+    if (status === "pending") {
+        return "⏳";
+    }
+    return "❌";
+}
+
+function OutputViewer({ status, content }) {
+    return (
+        <>
+            <label className="uk-form-label" htmlFor="content">
+                {status === "completed" ? "Resultat" : "Felmeddelande"}
+            </label>
+            <textarea
+                readOnly
+                rows={10}
+                id="content"
+                className={
+                    "uk-textarea uk-width-1-1 uk-resize-vertical" +
+                    (status === "completed" ? "" : " error")
+                }
+                disabled
+                value={content}
+            ></textarea>
+        </>
+    );
+}
+
+function Field({ label, value }) {
+    const id = useId(label);
+    return (
+        <>
+            <label className="uk-form-label" htmlFor={id}>
+                {label}:
+            </label>
+            <input
+                id={id}
+                className="uk-input uk-width-1-1"
+                disabled
+                value={value}
+            />
+        </>
+    );
+}
+
+function Signer({ member_id }) {
+    const [member, setMember] = useState(null);
+    const memberRef = useRef(Member.get(member_id));
+    useEffect(() => {
+        const m = memberRef.current;
+        return m.subscribe(() => {
+            setMember({
+                member_number: m.member_number,
+                firstname: m.firstname,
+                lastname: m.lastname,
+            });
+        });
+    }, [member_id]);
+
+    return (
+        <>
+            <label className="uk-form-label" htmlFor="signer">
+                Signerad av:
+            </label>
+            <input
+                id="signer"
+                className="uk-input uk-width-1-1"
+                disabled
+                value={
+                    member
+                        ? `#${member.member_number}: ${member.firstname} ${member.lastname}`
+                        : `${member_id}`
+                }
+            />
+        </>
+    );
+}
+
+function format_time(dt) {
+    const hour = String(dt.getHours()).padStart(2, "0");
+    const minute = String(dt.getMinutes()).padStart(2, "0");
+    return `${hour}:${minute}`;
+}
+
+function format_date(dt, include_time = false) {
+    const year = dt.getFullYear();
+    const month = String(dt.getMonth() + 1).padStart(2, "0");
+    const day = String(dt.getDate()).padStart(2, "0");
+
+    if (include_time) {
+        return `${year}-${month}-${day} ${format_time(dt)}`;
+    }
+
+    return `${year}-${month}-${day}`;
+}
+
+function Preview({ accounting_export }) {
+    return (
+        <>
+            <Signer member_id={accounting_export.signer_member_id} />
+            <Field
+                label="Status"
+                value={
+                    accounting_export.status === "completed"
+                        ? "Klar"
+                        : accounting_export.status === "pending"
+                          ? "Pågående"
+                          : "Exporten misslyckades"
+                }
+            />
+            <div className="uk-grid">
+                <div className="uk-width-1-2">
+                    <Field
+                        label="Start på period"
+                        value={format_date(
+                            new Date(accounting_export.start_date),
+                        )}
+                    />
+                </div>
+                <div className="uk-width-1-2">
+                    <Field
+                        label="Slut på period"
+                        value={format_date(
+                            new Date(accounting_export.end_date),
+                        )}
+                    />
+                </div>
+            </div>
+            <Field
+                label="Skapad"
+                value={format_date(
+                    new Date(accounting_export.created_at),
+                    true,
+                )}
+            />
+            <Field label="Aggregering" value={accounting_export.aggregation} />
+            {accounting_export.content && (
+                <OutputViewer
+                    status={accounting_export.status}
+                    content={accounting_export.content || ""}
+                />
+            )}
+        </>
+    );
+}
+
+function DownloadExportForm() {
+    const [exportedRanges, setExportedRanges] = useState([]);
+    const [selectedExport, setSelectedExport] = useState(null);
+    const id = useId();
+
+    const fetchExportedRanges = async () => {
+        console.log("Fetching exported ranges");
+        const response = await get({
+            url: "/webshop/accounting/export/",
+        });
+
+        const data_most_recent_first = response.data.sort((a, b) =>
+            new Date(a.updated_at).getTime() < new Date(b.updated_at).getTime()
+                ? 1
+                : -1,
+        );
+        const options = data_most_recent_first.map((d) => {
+            const start_date = format_date(new Date(d.start_date));
+            const end_date = format_date(new Date(d.end_date));
+            return {
+                label: `${d.id} ${status_to_emoji(d.status)} ${
+                    d.status
+                } (${start_date} - ${end_date})`,
+                value: d,
+            };
+        });
+        setExportedRanges(options);
+    };
+
+    useEffect(() => {
+        fetchExportedRanges();
+    }, []);
+
+    const can_download =
+        selectedExport && selectedExport.status === "completed";
+
+    return (
+        <form className="uk-form-stacked">
+            <fieldset className="uk-margin-top">
+                <legend className="uk-legend">
+                    Ladda ner exporterade perioder
+                </legend>
+                <label className="uk-form-label" htmlFor={id}>
+                    Exporter:
+                </label>
+                <div className="uk-grid">
+                    <div className="uk-width-2-3">
+                        <Select
+                            id={id}
+                            name="Export"
+                            className={"uk-width-1-1"}
+                            tabIndex={5}
+                            options={exportedRanges}
+                            selected={selectedExport}
+                            onChange={(selected) =>
+                                setSelectedExport(selected.value)
+                            }
+                            onMenuOpen={fetchExportedRanges}
+                        />
+                    </div>
+                    <div className="uk-width-1-3">
+                        <button
+                            className="uk-button uk-button-primary uk-width-1-1"
+                            role="button"
+                            disabled={!can_download}
+                            tabIndex={6}
+                            onClick={(event) => {
+                                event.preventDefault();
+                                downloadAccounting(selectedExport);
+                            }}
+                        >
+                            <Icon icon="cloud-download" /> Ladda ner
+                        </button>
+                    </div>
+                </div>
+                {selectedExport && (
+                    <Preview accounting_export={selectedExport} />
+                )}
+            </fieldset>
+        </form>
+    );
+}
+
+export default function AccountingExport() {
     return (
         <div>
             <h2>Exportera SIE-fil</h2>
@@ -81,58 +356,8 @@ export default function AccountingExport() {
                 På denna sida kan du exportera transaktioner för vald period
                 till en SIE-fil.
             </p>
-            <form className="uk-form-stacked">
-                <fieldset className="uk-margin-top">
-                    <legend className="uk-legend">
-                        Välj vilken period du vill exportera
-                    </legend>
-                    <label className="uk-form-label" htmlFor="year">
-                        År:
-                    </label>
-                    <Select
-                        id="year"
-                        name="year"
-                        className="uk-width-1-1"
-                        tabIndex={1}
-                        options={year_options}
-                        value={yearOption}
-                        onChange={(from_year) => setYearOption(from_year)}
-                    />
-                    <label className="uk-form-label" htmlFor="month">
-                        Månad:
-                    </label>
-                    <Select
-                        id="month"
-                        name="month"
-                        className="uk-width-1-1"
-                        tabIndex={2}
-                        options={month_options}
-                        value={monthOption}
-                        onChange={(month) => setMonthOption(month)}
-                    />
-                    <label className="uk-form-label" htmlFor="file_name">
-                        File name:
-                    </label>
-                    <input
-                        type="text"
-                        tabIndex={3}
-                        placeholder="Enter file name..."
-                        id="file_name"
-                        name="file_name"
-                        className="uk-input uk-width-1-1"
-                    />
-                    <button
-                        className="uk-button uk-button-primary uk-width-1-1"
-                        role="button"
-                        tabIndex={4}
-                        style={{ marginTop: "2px" }}
-                        disabled={loading}
-                        onClick={handleExportClick}
-                    >
-                        {loading ? "Laddar..." : "Exportera"}
-                    </button>
-                </fieldset>
-            </form>
+            <StartExportForm />
+            <DownloadExportForm />
         </div>
     );
 }

--- a/api/src/migrations/0030_accounting_export.sql
+++ b/api/src/migrations/0030_accounting_export.sql
@@ -1,0 +1,15 @@
+CREATE TABLE IF NOT EXISTS `webshop_accounting_exports` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `signer_member_id` int(10) unsigned NOT NULL,
+  `content` text CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci,
+  `status` enum('completed','pending','failed') CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL,
+  `aggregation` enum('year','month','day') CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL,
+  `start_date` date NOT NULL,
+  `end_date` date NOT NULL,
+  `created_at` datetime DEFAULT CURRENT_TIMESTAMP NOT NULL,
+  `updated_at` datetime DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+  `deleted_at` datetime DEFAULT NULL,
+  PRIMARY KEY (`id`),
+  KEY `webshop_accounting_export_member_id_index` (`signer_member_id`),
+  CONSTRAINT `webshop_accounting_export_member_id_foreign` FOREIGN KEY (`signer_member_id`) REFERENCES `membership_members` (`member_id`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;

--- a/api/src/service/config.py
+++ b/api/src/service/config.py
@@ -94,6 +94,9 @@ def get_admin_url(path: str) -> str:
 
 def get_makerspace_local_timezone() -> ZoneInfo:
     zone_str = config.get("MAKERSPACE_LOCAL_TIMEZONE")
+    if zone_str is None:
+        raise NameError("Variable MAKERSPACE_LOCAL_TIMEZONE not set in .env")
+
     try:
         zone = ZoneInfo(zone_str)
     except ZoneInfoNotFoundError as e:

--- a/api/src/service/config.py
+++ b/api/src/service/config.py
@@ -36,6 +36,7 @@ default = Dict(
         ELKS46_API_USER="",
         ELKS46_API_KEY="",
         HOST_PUBLIC="",
+        MAKERSPACE_LOCAL_TIMEZONE="Europe/Stockholm",
         STRIPE_PRIVATE_KEY=None,
         STRIPE_SIGNING_SECRET=None,
         STRIPE_CURRENCY=None,

--- a/api/src/services.py
+++ b/api/src/services.py
@@ -10,6 +10,7 @@ import multiaccess
 import multiaccessy
 import quiz
 import shop
+import shop.accounting
 
 Instance = namedtuple("Instance", "path,service")
 
@@ -18,6 +19,7 @@ services = (
     Instance(path="", service=core.service),
     Instance(path="/membership", service=membership.service),
     Instance(path="/webshop", service=shop.service),
+    Instance(path="/webshop/accounting", service=shop.accounting.service),
     Instance(path="/member", service=member.service),
     Instance(path="/messages", service=messages.service),
     Instance(path="/statistics", service=statistics.service),

--- a/api/src/shop/accounting/__init__.py
+++ b/api/src/shop/accounting/__init__.py
@@ -1,0 +1,7 @@
+from service.config import config
+from service.internal_service import InternalService
+
+service = InternalService(name="/webshop/accounting")
+
+# This import is needed to register the routes
+import shop.accounting.views

--- a/api/src/shop/accounting/accounting.py
+++ b/api/src/shop/accounting/accounting.py
@@ -7,7 +7,7 @@ from typing import Dict, List, Tuple
 
 from basic_types.enums import AccountingEntryType
 from service.db import db_session
-from service.error import InternalServerError
+
 from shop.completed_payment import CompletedPayment
 from shop.models import (
     Product,
@@ -19,6 +19,10 @@ from shop.models import (
 )
 
 logger = getLogger("makeradmin")
+
+
+class AccountingError(Exception):
+    pass
 
 
 @dataclass()
@@ -79,11 +83,11 @@ class ProductToAccountCostCenter:
                 account = product_info.account if product_info.account is not None else None
                 cost_center = product_info.cost_center if product_info.cost_center is not None else None
                 if account is None and cost_center is None:
-                    raise InternalServerError(
+                    raise AccountingError(
                         f"Product with id {product.id} has accounting with both account and cost center as None. At least one must be set to a value."
                     )
                 if product_info.fraction <= 0 or product_info.fraction > 100:
-                    raise InternalServerError(
+                    raise AccountingError(
                         f"Product with id {product.id} has accounting with id {product_info.id} with fraction {product_info.fraction} not in range [1, 100]"
                     )
 
@@ -101,11 +105,11 @@ class ProductToAccountCostCenter:
 
             for key in fraction_sums:
                 if not entry_type_found[key]:
-                    raise InternalServerError(
+                    raise AccountingError(
                         f"Product with id {product.id} named {product.name} has no accounting information for {key.value}"
                     )
                 if fraction_sums[key] != 100:
-                    raise InternalServerError(
+                    raise AccountingError(
                         f"Product with id {product.id} named {product.name} has accounting type {key.value} "
                         + f"with fraction weights not adding up to 100, was {fraction_sums[key]}",
                     )
@@ -119,7 +123,7 @@ class ProductToAccountCostCenter:
 def diff_transactions_and_completed_payments(
     transactions: List[Transaction], completed_payments: Dict[int, CompletedPayment]
 ) -> List[Tuple[Transaction | None, CompletedPayment | None]]:
-    unmatched_data: List[Tuple[Transaction | None, CompletedPayment]] = []
+    unmatched_data: List[Tuple[Transaction | None, CompletedPayment | None]] = []
     completed_payments = completed_payments.copy()  # TODO improve this and remove pop
 
     for transaction in transactions:
@@ -179,7 +183,7 @@ def split_transactions_over_accounts(
         for content in transaction.contents:
             product_accounting = product_to_accounting.get_account_cost_center(content.product_id)
             if not product_accounting:
-                raise InternalServerError(f"Product with id {content.product_id} has no accounting information")
+                raise AccountingError(f"Product with id {content.product_id} has no accounting information")
 
             amounts_added: Dict[AccountingEntryType, Decimal] = {type: Decimal(0) for type in AccountingEntryType}
             content_amount_decimal = Decimal(

--- a/api/src/shop/accounting/export.py
+++ b/api/src/shop/accounting/export.py
@@ -1,19 +1,22 @@
 from datetime import datetime
 from logging import getLogger
-from typing import Dict, List, Tuple
+from typing import Dict, List
 
 from basic_types.enums import AccountingEntryType
 from basic_types.time_period import TimePeriod
 from membership.models import Member
+from service.config import get_makerspace_local_timezone
 from service.db import db_session
 from service.error import InternalServerError
+from zoneinfo import ZoneInfo
+
 from shop.accounting.accounting import (
-    RoundingError,
-    RoundingErrorSource,
+    AccountingError,
     TransactionWithAccounting,
     diff_transactions_and_completed_payments,
     split_transactions_over_accounts,
 )
+from shop.accounting.models import AccountingExport, Aggregation, Status
 from shop.accounting.sie_file import get_sie_string
 from shop.accounting.verification import create_verificatons
 from shop.completed_payment import CompletedPayment, get_completed_payments_from_stripe
@@ -23,7 +26,6 @@ from shop.models import (
     TransactionContent,
     TransactionCostCenter,
 )
-from zoneinfo import ZoneInfo
 
 logger = getLogger("makeradmin")
 
@@ -49,10 +51,7 @@ def transaction_fees_to_transaction_with_accounting(
     return amounts
 
 
-def export_accounting(start_date: datetime, end_date: datetime, group_by_period: TimePeriod, member_id: int) -> str:
-    signer = db_session.query(Member).filter(Member.member_id == member_id).one_or_none()
-    if signer is None:
-        raise InternalServerError(f"Member with id {member_id} not found")
+def export_accounting(start_date: datetime, end_date: datetime, aggregation: Aggregation, signer: Member) -> str:
     utc_zone = ZoneInfo("UTC")
     logger.info(
         f"Exporting accounting from {start_date} ({start_date.astimezone(utc_zone)}) to {end_date} ({end_date.astimezone(utc_zone)}) with signer member number {signer.member_number}"
@@ -77,15 +76,38 @@ def export_accounting(start_date: datetime, end_date: datetime, group_by_period:
 
     diff = diff_transactions_and_completed_payments(transactions, completed_payments)
     if len(diff) > 0:
-        logger.warning(f"Transactions and completed payments do not match, {diff}")
-        raise InternalServerError(f"Transactions and completed payments do not match, {diff}")
+        raise AccountingError(f"Transactions and completed payments do not match, {diff}")
 
     transactions_with_accounting, rounding_errors = split_transactions_over_accounts(transactions, completed_payments)
 
     transaction_fees = transaction_fees_to_transaction_with_accounting(completed_payments)
     transactions_with_accounting.extend(transaction_fees)
+    group_by_period = TimePeriod(aggregation.value)  # aggregation is in fact a subset
 
     verifications = create_verificatons(transactions_with_accounting, group_by_period)
     logger.info(f"Verifications: {verifications}")
 
     return get_sie_string(verifications, start_date, end_date, f"{signer.firstname} {signer.lastname}")
+
+
+def do_export(export: AccountingExport) -> None:
+    if export.status not in (Status.pending, Status.failed):
+        raise ValueError(f"Export with id {export.id} is not pending or failed")
+
+    export.content = None
+    export.status = Status.pending
+
+    zone = get_makerspace_local_timezone()
+    start_dt = datetime.combine(export.start_date, datetime.min.time(), tzinfo=zone)
+    end_dt = datetime.combine(export.end_date, datetime.min.time(), tzinfo=zone)
+
+    try:
+        accounting_info = export_accounting(start_dt, end_dt, export.aggregation, export.signer)
+        export.content = accounting_info
+        export.status = Status.completed
+    except Exception as e:
+        export.content = str(e)
+        export.status = Status.failed
+        logger.exception(f"Failed to export accounting for {export}")
+
+    db_session.commit()

--- a/api/src/shop/accounting/models.py
+++ b/api/src/shop/accounting/models.py
@@ -1,0 +1,59 @@
+import enum
+from datetime import date, datetime
+from typing import Any
+
+from membership.models import Member
+from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship
+from sqlalchemy.schema import ForeignKey
+from sqlalchemy.sql import func
+from sqlalchemy.types import DateTime, Integer, Text
+
+
+class Base(DeclarativeBase):
+    pass
+
+
+class Status(enum.Enum):
+    completed = "completed"
+    pending = "pending"
+    failed = "failed"
+
+
+class Aggregation(enum.Enum):
+    year = "year"
+    month = "month"
+    day = "day"
+
+
+class AccountingExport(Base):
+    __tablename__ = "webshop_accounting_exports"
+
+    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
+    signer_member_id: Mapped[int] = mapped_column(Integer, ForeignKey(Member.member_id))
+    content: Mapped[str | None] = mapped_column(Text)
+    status: Mapped[Status] = mapped_column(default=Status.pending)
+    aggregation: Mapped[Aggregation] = mapped_column(default=Aggregation.month)
+    start_date: Mapped[date]
+    end_date: Mapped[date]
+    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
+    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
+    deleted_at: Mapped[datetime | None]
+
+    signer: Mapped[Member] = relationship()
+
+    def __repr__(self) -> str:
+        return f"AccountingFiles(id={self.id}, status={self.status}, start_date={self.start_date}, end_date={self.end_date}, aggregation={self.aggregation})"
+
+    def to_dict(self) -> dict[str, Any]:
+        return {
+            "id": self.id,
+            "content": self.content,
+            "status": self.status.value,
+            "aggregation": self.aggregation.value,
+            "start_date": self.start_date,
+            "end_date": self.end_date,
+            "created_at": self.created_at,
+            "updated_at": self.updated_at,
+            "deleted_at": self.deleted_at,
+            "signer_member_id": self.signer_member_id,
+        }

--- a/api/src/shop/accounting/sie_file.py
+++ b/api/src/shop/accounting/sie_file.py
@@ -5,7 +5,8 @@ from typing import Dict, List, Tuple
 
 from basic_types.enums import AccountingEntryType
 from membership.models import Member
-from shop.accounting.accounting import TransactionAccount, TransactionCostCenter
+
+from shop.accounting.accounting import AccountingError, TransactionAccount, TransactionCostCenter
 from shop.accounting.verification import Verification
 
 logger = getLogger("makeradmin")
@@ -23,6 +24,10 @@ HEADER_TEMPLATE = """#FLAGGA 0
 #VALUTA SEK
 #DIM 1 "Kostnadställe"
 """
+
+
+class SieExportError(AccountingError):
+    pass
 
 
 def period_to_date_format(period: str) -> str:
@@ -65,7 +70,7 @@ def get_account_header(verifications: List[Verification]) -> List[str]:
             if account not in seen_accounts:
                 seen_accounts.add(account)
                 if account is None:
-                    raise ValueError("Account cannot be None for SIE export.")
+                    raise SieExportError("Account cannot be None for SIE export.")
                 accounts_header.append(f'#KONTO {account.account} "{account.description}"')
 
     return accounts_header
@@ -100,7 +105,7 @@ def convert_to_sie_format(verifications: List[Verification]) -> List[str]:
             account = accounting_key[0]
             cost_center = accounting_key[1]
             if account is None:
-                raise ValueError("Account cannot be None for SIE export.")
+                raise SieExportError("Account cannot be None for SIE export.")
             sie_amount_adjusted = (
                 -amount if verification.types[accounting_key] == AccountingEntryType.CREDIT else amount
             )

--- a/api/src/shop/accounting/test/accounting_test.py
+++ b/api/src/shop/accounting/test/accounting_test.py
@@ -15,6 +15,7 @@ from service.db import db_session
 from service.error import InternalServerError
 from shop.accounting.accounting import (
     AccountCostCenter,
+    AccountingError,
     ProductToAccountCostCenter,
     RoundingErrorSource,
     TransactionWithAccounting,
@@ -264,7 +265,7 @@ class ProductToAccountCostCenterTest(FlaskTestBase):
             type=AccountingEntryType.CREDIT.value,
             fraction=0,
         )
-        with self.assertRaises(InternalServerError) as context:
+        with self.assertRaises(AccountingError) as context:
             ProductToAccountCostCenter()
         self.assertTrue("not in range [1, 100]" in str(context.exception))
 
@@ -277,7 +278,7 @@ class ProductToAccountCostCenterTest(FlaskTestBase):
             type=AccountingEntryType.CREDIT.value,
             fraction=101,
         )
-        with self.assertRaises(InternalServerError) as context:
+        with self.assertRaises(AccountingError) as context:
             ProductToAccountCostCenter()
         self.assertTrue("not in range [1, 100]" in str(context.exception))
 
@@ -290,7 +291,7 @@ class ProductToAccountCostCenterTest(FlaskTestBase):
             type=AccountingEntryType.CREDIT.value,
             fraction=1,
         )
-        with self.assertRaises(InternalServerError) as context:
+        with self.assertRaises(AccountingError) as context:
             ProductToAccountCostCenter()
         self.assertTrue("fraction weights not adding up to 100" in str(context.exception))
 
@@ -301,7 +302,7 @@ class ProductToAccountCostCenterTest(FlaskTestBase):
                 acc.fraction = acc.fraction + 1
                 break
         db_session.commit()
-        with self.assertRaises(InternalServerError) as context:
+        with self.assertRaises(AccountingError) as context:
             ProductToAccountCostCenter()
         self.assertTrue("fraction weights not adding up to 100" in str(context.exception))
 
@@ -312,7 +313,7 @@ class ProductToAccountCostCenterTest(FlaskTestBase):
                 acc.fraction = acc.fraction - 1
                 break
         db_session.commit()
-        with self.assertRaises(InternalServerError) as context:
+        with self.assertRaises(AccountingError) as context:
             ProductToAccountCostCenter()
         self.assertTrue("fraction weights not adding up to 100" in str(context.exception))
 
@@ -325,7 +326,7 @@ class ProductToAccountCostCenterTest(FlaskTestBase):
             type=AccountingEntryType.CREDIT.value,
             fraction=100,
         )
-        with self.assertRaises(InternalServerError) as context:
+        with self.assertRaises(AccountingError) as context:
             ProductToAccountCostCenter()
         self.assertTrue("both account and cost center as None" in str(context.exception))
 

--- a/api/src/shop/accounting/test/export_test.py
+++ b/api/src/shop/accounting/test/export_test.py
@@ -1,24 +1,23 @@
-from datetime import date, datetime, timezone
+from datetime import datetime, timezone
 from decimal import Decimal
 from logging import getLogger
-from typing import Dict, List, Tuple
-from unittest import TestCase
+from typing import Dict, List
 from unittest.mock import Mock, patch
 
 import core
 import membership
-import pytest
 import shop
+import shop.accounting
 from basic_types.enums import AccountingEntryType
-from basic_types.time_period import TimePeriod, date_to_period
+from basic_types.time_period import TimePeriod
 from membership.models import Member
 from service.db import db_session
-from shop.accounting.accounting import RoundingError, RoundingErrorSource
 from shop.accounting.export import (
+    do_export,
     export_accounting,
     transaction_fees_to_transaction_with_accounting,
 )
-from shop.accounting.verification import Verification
+from shop.accounting.models import AccountingExport, Aggregation, Status
 from shop.completed_payment import CompletedPayment
 from shop.models import (
     Product,
@@ -34,7 +33,12 @@ from test_aid.test_base import FlaskTestBase
 logger = getLogger("makeradmin")
 
 
-class ExportTest(TestCase):
+class ExportTest(FlaskTestBase):
+    models = [core.models, membership.models, shop.models, shop.accounting.models]
+
+    def setUp(self) -> None:
+        self.member = self.db.create_member(firstname="Test", lastname="Testsson")
+
     def test_fees_to_transaction_with_accounting(self) -> None:
         completed_payments: Dict[int, CompletedPayment] = {}
         num_payments = 10
@@ -55,6 +59,27 @@ class ExportTest(TestCase):
             self.assertEqual(transaction_fees[i].amount, true_fees[i])
             assert transaction_fees[i].account.account == "6573"
             assert transaction_fees[i].cost_center.cost_center == "FG"
+
+    @patch("shop.accounting.export.export_accounting")
+    def test_create_accounting_file(self, export_accounting: Mock) -> None:
+        start_date = datetime(2023, 1, 1)
+        end_date = datetime(2023, 12, 31)
+
+        accounting_info = "accounting string"
+        export_accounting.return_value = accounting_info
+        accounting_export = AccountingExport(
+            signer_member_id=self.member.member_id,
+            aggregation=Aggregation.month,
+            start_date=start_date,
+            end_date=end_date,
+        )
+        db_session.add(accounting_export)
+        db_session.commit()
+
+        do_export(accounting_export)
+
+        self.assertEqual(accounting_export.status, Status.completed)
+        self.assertEqual(accounting_export.content, accounting_info)
 
 
 class AccountingExportWithStripeMockTest(FlaskTestBase):
@@ -168,7 +193,7 @@ class AccountingExportWithStripeMockTest(FlaskTestBase):
 
         get_payments_from_stripe.return_value = self.completed_payments
 
-        sie_str = export_accounting(start_date, end_date, TimePeriod.Month, self.member.member_id)
+        sie_str = export_accounting(start_date, end_date, Aggregation.month, self.member)
         sie_rows = sie_str.split("\n")
 
         while True:

--- a/api/src/shop/accounting/verification.py
+++ b/api/src/shop/accounting/verification.py
@@ -7,7 +7,13 @@ from basic_types.enums import AccountingEntryType
 from basic_types.time_period import TimePeriod, date_to_period
 from service.db import db_session
 from service.error import InternalServerError
-from shop.accounting.accounting import TransactionAccount, TransactionCostCenter, TransactionWithAccounting
+
+from shop.accounting.accounting import (
+    AccountingError,
+    TransactionAccount,
+    TransactionCostCenter,
+    TransactionWithAccounting,
+)
 
 logger = getLogger("makeradmin")
 
@@ -30,7 +36,7 @@ def create_verificatons(
         if year_month in verifications:
             if inner_key in verifications[year_month].amounts:
                 if verifications[year_month].types[inner_key] != transaction.type:
-                    raise InternalServerError(
+                    raise AccountingError(
                         f"Multiple transactions with different types for the same account, {transaction.account}, and cost center, {transaction.cost_center}, in the same period {year_month}."
                     )
                 verifications[year_month].amounts[inner_key] += transaction.amount

--- a/api/src/shop/accounting/views.py
+++ b/api/src/shop/accounting/views.py
@@ -1,0 +1,60 @@
+import multiprocessing
+from datetime import date
+from typing import Any
+
+from dateutil.relativedelta import relativedelta
+from flask import g
+from membership.models import Member
+from service.api_definition import GET, POST, WEBSHOP
+from service.db import db_session
+from service.error import InternalServerError
+from sqlalchemy import select
+
+from shop.accounting import service
+from shop.accounting.export import do_export
+from shop.accounting.models import AccountingExport, Aggregation
+
+
+@service.route("/export/<int:year>/<int:month>", method=POST, permission=WEBSHOP)
+def start_accounting_export_route(year: int, month: int) -> int:
+    start_date = date(year=year, month=month, day=1)
+    end_date = start_date + relativedelta(months=1)
+
+    signer_member_id: int = g.user_id
+
+    signer = db_session.get(Member, signer_member_id)
+    if signer is None:
+        raise ValueError(f"Member with id {signer_member_id} not found")
+
+    e = AccountingExport(
+        signer_member_id=signer_member_id,
+        aggregation=Aggregation.month,
+        start_date=start_date,
+        end_date=end_date,
+    )
+    db_session.add(e)
+    db_session.flush()
+
+    created_id = e.id
+    assert created_id is not None
+
+    db_session.commit()
+
+    job = multiprocessing.Process(target=do_export, args=(e,))
+    job.start()
+
+    return created_id
+
+
+@service.route("/export/", method=GET, permission=WEBSHOP)
+def list_accounting_files_route() -> list[dict[str, Any]]:
+    return [f.to_dict() for f in db_session.execute(select(AccountingExport)).scalars()]
+
+
+@service.route("/export/<int:export_id>", method=GET, permission=WEBSHOP)
+def get_export_route(export_id: int) -> dict[str, Any]:
+    export = db_session.get(AccountingExport, export_id)
+    if export is None:
+        raise InternalServerError(f"Accounting file with id {export_id} not found")
+
+    return export.to_dict()

--- a/api/src/shop/completed_payment.py
+++ b/api/src/shop/completed_payment.py
@@ -47,12 +47,12 @@ def convert_completed_stripe_charges_to_payments(
         if not charge.paid:
             continue
 
-        assert charge.balance_transaction is not None
-
-        if not MakerspaceMetadataKeys.TRANSACTION_IDS.value in charge.metadata:
-            # This is a temporary fix until we done the accounting for all the old stuff. Can be removed after 2025-01-01
-            intent = retry(lambda: PaymentIntent.retrieve(charge.payment_intent))
-            id = int(intent.metadata[MakerspaceMetadataKeys.TRANSACTION_IDS.value])
+        if charge.balance_transaction is None:
+            logger.error(f"Missing balance transaction in stripe charge, {charge.id}")
+            raise BadRequest(f"Missing balance transaction in stripe charge, {charge.id}")
+        elif not MakerspaceMetadataKeys.TRANSACTION_IDS.value in charge.metadata:
+            logger.error(f"Missing transaction id in stripe charge metadata, {charge.id}")
+            raise BadRequest(f"Missing transaction id in stripe charge metadata, {charge.id}")
         else:
             id = int(charge.metadata[MakerspaceMetadataKeys.TRANSACTION_IDS.value])
 

--- a/api/src/shop/models.py
+++ b/api/src/shop/models.py
@@ -9,7 +9,6 @@ from service.error import UnprocessableEntity
 from sqlalchemy import (
     JSON,
     Boolean,
-    Column,
     DateTime,
     Enum,
     ForeignKey,

--- a/api/src/shop/views.py
+++ b/api/src/shop/views.py
@@ -1,23 +1,17 @@
-from base64 import b64encode
-from dataclasses import dataclass
-from datetime import datetime, timedelta, timezone
+from datetime import datetime
 from logging import getLogger
-from typing import Any, Optional
+from typing import Any
 
 from basic_types.enums import PriceLevel
-from basic_types.time_period import TimePeriod
-from dateutil.relativedelta import relativedelta
 from flask import Response, g, make_response, request, send_file
 from multiaccessy.invite import AccessyInvitePreconditionFailed, ensure_accessy_labaccess
-from service.api_definition import DELETE, GET, MEMBER_EDIT, POST, PUBLIC, USER, WEBSHOP, WEBSHOP_EDIT, Arg
-from service.config import get_makerspace_local_timezone
+from service.api_definition import DELETE, GET, MEMBER_EDIT, POST, PUBLIC, USER, WEBSHOP, WEBSHOP_EDIT
 from service.db import db_session
 from service.entity import OrmSingeRelation, OrmSingleSingleRelation
-from service.error import BadRequest, InternalServerError, PreconditionFailed
+from service.error import BadRequest, PreconditionFailed
 from sqlalchemy.exc import NoResultFound
 
 from shop import service
-from shop.accounting.export import export_accounting
 from shop.entities import (
     category_entity,
     gift_card_content_entity,
@@ -324,18 +318,6 @@ def register_route() -> Any:
 @service.route("/stripe_callback", method=POST, permission=PUBLIC, commit_on_error=True)
 def stripe_callback_route() -> None:
     stripe_callback(request.data, request.headers)
-
-
-@service.route("/download-accounting-file/<int:year>/<int:month>", method=GET, permission=WEBSHOP)
-def download_accounting_file_route(year: int, month: int) -> str:
-    zone = get_makerspace_local_timezone()
-    start_date = datetime(year=year, month=month, day=1, hour=0, minute=0, second=0, microsecond=0, tzinfo=zone)
-    end_data = datetime(
-        year=year, month=month, day=1, hour=0, minute=0, second=0, microsecond=0, tzinfo=zone
-    ) + relativedelta(months=1)
-    return b64encode(export_accounting(start_date, end_data, TimePeriod.Month, g.user_id).encode("cp437")).decode(
-        "ascii"
-    )
 
 
 @service.route("/product/<int:product_id>/sales", method=GET, permission=WEBSHOP)

